만국박람회 [STEP 2] 수꿍, yeton

yeeton37 opened the pull request:
안녕하세요, 쿠마 @leejun6694! 🤗🤗 

만국박람회 프로젝트 step2를 함께 진행한 수꿍과 yeton입니다.
아래에 프로젝트 구현 설명과 그 과정에서 고민했던 점과 의문점에 대해 적어보았습니다.
<br/>

## 실행 화면(기능 설명)

|파리만국박람회 소개 화면|한국 출품작 소개 화면|출품작 상세 소개 화면|
|:--:|:--:|:--:|
|![EXPOInformationViewController](https://user-images.githubusercontent.com/99063327/174227640-f8d1a00b-957a-449e-b48b-f0b9e4d0aa2c.gif)|![KoreaEntryViewController](https://user-images.githubusercontent.com/99063327/174229792-2c3bb96c-1609-4bf2-b065-7957198ffe64.gif)|![KoreaEntryDetailsViewController](https://user-images.githubusercontent.com/99063327/174230832-9e656db0-add6-44b7-bf48-84cf62fa37dd.gif)|

### 1. 파리만국박람회 소개 화면(EXPOInformationViewController)

- `ScrollView` 내에 파리 만국박람회의 `title`, `image`, `visitors`, `location`, `duration`, `description`에 관한 `label` 및 `imageView`를 생성하여 스크롤을 통해 원하는 내용을 볼 수 있게 하였다.
- `view`의 최하단에 '한국의 출품작 보러가기'라는 이름을 가진 버튼을 생성하여, `KoreaEntryViewController`로 화면 전환(`push`) 기능을 수행하였다.
<br/>

### 2. 한국 출품작 소개 화면(KoreaEntryViewController)

- `TableView`를 생성하여 각각의 셀에 한국의 출품작의 `image`, `name`, `short_desc`을 표시하였다.
- 셀을 클릭하면 해당 출품작에 관한 `image`와 `description`을 표시하는 세번째 화면으로 화면전환(`segue`) 기능을 수행하였다.
<br/>

### 3. 출품작 상세 소개 화면(KoreaEntryDetailsViewController)

- `ScrollView`내 `image`와 `description`에 관한 `label` 및 `ImageView`를 생성하여 스크롤을 통해 원하는 내용을 볼 수 있게 하였다.
- `NavigationBar`의 `BackButton`을 통하여 이전 화면인 `KoreaEntryViewController`로 돌아가는 기능도 추가하였다.

<br/>

## 고민했던 점

### 주제
### 1. xib 파일 ViewController와 연결
- 파일을 분리해 Cell과 Cell의 xib 파일을 만들어 ViewController에서 사용해주려했는데 cell이 나타나지 않는 현상이 발생함
- 그 이유를 찾아보니, xib 파일은 따로 뷰에 register를 해줘야 사용할 수 있음을 알게되었다.
- 아래와 같이 xib 파일을 등록해주니 cell이 정상적으로 뷰에 나타나는 것을 확인할 수 있었다.
```swift
let nibName = UINib(nibName: "CustomCell", bundle: nil)
        tableView.register(nibName, forCellReuseIdentifier: "CustomCell")
```

### 2. Navigation Bar 숨기기
- 뷰가 화면에 보일 때 Navigation Bar를 숨기고 다른 창으로 넘어갈 때 보이게 하는 방법에 대해 고민했다.
- 이는 아래와 같이 View Life Cycle 메소드를 활용하여 해결할 수 있었다.
```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    self.navigationController?.isNavigationBarHidden = true
}
    
override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
    self.navigationController?.isNavigationBarHidden = false
}
```

### 3. 스크롤 뷰의 오토레이아웃
- 스크롤뷰에 추가한 `contentView`의 오토레이아웃을 스크롤뷰의 `width`와 `Equal Width`로 설정해주니 오류가 발생했다.
- 찾아보니, `contentView` 또한 실질적인 사이즈가 필요하기 때문에 `content`의 `width`와 `height`가 필요하다는 점을 알게되었다.
- 또한 추가된 `contentView`를 스크롤 뷰의 `Content Layout Guide`에 오토 레이아웃을 사용해 꽉 채워주고, `contentView`를 `control`을 누른상태로 `Content Layout Guide`에 드래그하여 오토 레이아웃을 줄 수 있다. 이 방법으로 `leading`, `top`, `trailing`, `bottom anchor` 모두 `constraint`를 적용해주어야 함을 알게되었다.

[[스크롤 뷰 오토레이아웃 주는 순서]](https://corykim0829.github.io/ios/UIScrollView-with-storyboard/#)
1. 스크롤 뷰 오토레이아웃 잡아주기
2. `ContentView` 추가하기
3. `Content View` 사이즈 잡아주기
4. `Frame Layout` 잡아주기
5. `Width`/`Height` 잡아주기


<br/>

## 의문점

### 주제
### 1. 네비게이션 타이틀이 안바뀌는 현상 

- 네비게이션 바에 `title`과 `backButtonTitle`을 설정해주는 방법으로 다음의 두 방법을 발견하였습니다.

```swift
// 1번 방법
navigationItem.title = "한국의 출품작"
navigationItem.backButtonTitle = "메인"

// 2번 방법
navigationController?.navigationBar.topItem?.title = "한국의 출품작"
navigationController?.navigationBar.topItem?.backButtonTitle = "메인"
```
|두번째 VC내 1번 방법 사용|두번째 VC내 2번 방법 사용|첫번째 VC와 두번째 VC 내 1번 방법 사용|
|:--:|:--:|:--:|
|<img width="338" alt="secondVC 내 navigationController" src="https://user-images.githubusercontent.com/99063327/174234793-02f81c9a-ee30-4732-98cc-aa82ebde61d8.png">|<img width="336" alt="secondVC 내 navigationItem" src="https://user-images.githubusercontent.com/99063327/174234797-c0d15197-10d9-4f7b-8edb-610ff542c11f.png">|<img width="337" alt="firstVC 내 backbutton 및 secondVC 내 navigationItem" src="https://user-images.githubusercontent.com/99063327/174234783-0552fbe2-64b8-4753-bd70-cee9a221e420.png">|
|`KoreaEntryViewController` 내의 `viewDidLoad` 함수에서 첫번째 방법을 사용해보니 위와 같이 백버튼은 원하는 네이밍이 잘 표시되었으나, 네비게이션 타이틀은 표시되지 않았습니다.|`KoreaEntryViewController` 내의 `viewDidLoad` 함수에서 두번째 방법을 사용해보니 이버에는 위와 같이 네비게이션 타이틀은 원하는 네이밍이 잘 표시되었으나, 백 버튼의 네이밍이 원하는 대로 표시되지 않았습니다.|이번에는 첫번째 방법을 사용하되, `backButtonTitle`은 첫번째 `VC`인 `EXPOInformationViewController`에서 생성하고, `navigation title`은 두번째 VC인 `KoreaEntryViewController`에서 생성하였습니다. <br/> 그러면 위와같이 네이밍이 원하는 바와 같이 나올 수 있었습니다.|



- 문제점을 해결할 수는 있었으나, 왜 그렇게 구현되는지 원리를 제대로 이해하지는 못했습니다. 저희가 이해한 바로는 첫번째 방법과 두번째 방법은 충분히 서로를 대체할 수 있는 함수로 인식하였는데, 같은 위치에 메서드를 호출하여도 다른 결과를 보였습니다. 이에 그 원리를 질문드리고 싶습니다.

- 또한, `backButtonTitle`은 두번째 `VC`가 아닌, 첫번째 `VC`에서 설정해줘야 원하는 네이밍을 생성할 수 있었는데, 이렇게 작용하는 근거가 무엇인지 궁금합니다. 


### 2. 이미지 사이즈 dynamic하게 변경하는 법
![](https://i.imgur.com/fq7mou3.png)

- 아래 코드에서 `defaultContentConfiguration` 메서드를 사용해 위의 `cell`을 만들어주었고, `cell` 안의 `content`의 `imageProperties`의 사이즈를 50으로 정해주었습니다.

- 그런데 사이즈를 50으로 `static`하게 지정해주는 것이 아니라 기기마다 알맞게 적용되도록 `dynamic`하게 설정하고 싶었습니다

- 그런데 메서드나 속성을 찾아봐도 `dynamic`하게 이미지 사이즈를 설정해주는 것에 대해 나와있지 않아서 그냥 이대로 설정해주었습니다.

- `defaultContentConfiguration` 메서드를 사용해 만들어준 셀 내부의 이미지를 옆의 `text`와 `secondaryText`와의 관계를 통해 오토레이아웃을 지정해주는 방법이 따로 있는지 질문드리고 싶습니다.

- 코드
```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "entryCell", for: indexPath)
    var content = cell.defaultContentConfiguration()
        
    content.image = UIImage(named: koreaEntry[indexPath.row].imageName)
    content.imageProperties.maximumSize.width = 50 [ 여기 !]
    content.imageProperties.maximumSize.height = 50 [ 여기 !]
        
    cell.contentConfiguration = content
      
    return cell
    }
```

### 3. JSONData 구조체의 parse 메서드

- 이번 만국박람회 프로젝트에서 `JSONData`를 `parse`하는 작업은 총 두 번 이루어졌습니다.
- 처음은 첫번째 화면에서 만국박람회 포스터 내의 이름, 이미지, 개최지, 기한, 설명 등에 대한 데이터를 가져오기 위해 이루어졌고, 다음으로 두번째 화면에서 각각의 한국 출품작에 대한 이미지, 타이틀, `short_desc`을 가져오기 위하여 파싱이 이루어졌습니다.
- 이 둘의 코드를 비교해본 결과, `NSDataAsset`의 `name`과 `decoding`할 타입을 제외하고는 코드를 동일하게 사용하고 있다고 판단하여, 다음과 같이 코드를 작성하였습니다.
- 코드
```swift
struct Parser {
    static func parse<T>(name: String, parsedItems: [T]) -> [T]? where T: Decodable {
        let jsonDecoder: JSONDecoder = JSONDecoder()
        
        let parsedItemsType = type(of: parsedItems)

        guard let dataAsset: NSDataAsset = NSDataAsset(name: name) else {
            return nil
        }

        do {
            let data = try jsonDecoder.decode(parsedItemsType.self, from: dataAsset.data)
            return data
        } catch {
            self.handleError(error)
            return nil
        }
    }
}
```
- 그러나, 이 `parse` 메서드는 두번째 `VC`에서 파싱을 진행하였을 때는 정상적으로 작동하였으나, 첫번째 `VC`에서는 정상적으로 작동하지 않았습니다.
- 그 근거를 추측해본 결과, `JSON Data` 파일을 뜯어보니 두번째 `VC`에서 사용하는 `Items` 데이터는 배열안에 객체들이 들어있는 형태로, 위의 메서드를 적용할 수 있었으나, 첫번째 `VC`에서 사용하는 `exposition_universelle_1900` 데이터는 배열이 있지 않고, 하나의 객체만이 들어있었습니다.
- 첫번째 `VC`에서도 위의 `parse` 메서드를 쓰기 위하여 아래와 같이 JSON Data를 담을 모델의 타입을 변경하였으나, 근본적인 `JSON` 데이터 타입의 문제 때문인지 해당 배열 내에 값이 온전히 할당되지 않았습니다.
```swift=
// 기존
var expositionUniverselle: ExpositionUniverselle? = nil

// 변경
var expositionUniverselle: [ExpositionUniverselle] = []
```
- 그래서 일단, 이를 해결하기 위하여 오버로딩 방식을 통하여, 타입이 `T`인 경우와 `[T]`인 경우로 파라미터 및 반환값을 변경하여 적용하였는데, 위의 하나의 메서드만을 이용하여 할 수 있는 방법은 정녕 없는지 질문드리고 싶습니다.
- 추가로, 해당 `parse` 메서드 내에서 실제 파라미터는 필요없지만 제너릭의 타입을 설정하기 위하여 `parsedItems`에서 파라미터를 받고, 그 파라미터에 `type(of:)` 메서드를 사용하여 메타타입의 값을 사용했습니다.
- 이렇게 내부 메서드를 구현하는 게 과연 옳은 것인지 또한 묻고 싶습니다.


### 4. 프로젝트에서 MVC 디자인 패턴 적용
- 현재 뷰에 보이고있는 라벨과 텍스트들을 따로 `UIView`를 상속한 `View` 타입으로 이동시켜준 뒤 `Controller`에서 `View`를 가져오는 방법을 사용하고자 했으나 충분히 알아보고 적용하고자 이번 step2에서는 아직 적용하지 못했습니다.
- 따라서 다음 step3부터는 `View`와 `Controller`를 분리해서 적용하려고 하는데, 보통 이렇게 `View`와 `Controller`를 나누는 방법을 선호하는지 궁금합니다.

### 5. 파리만국박람회 Title 개행 관련

- 첫번째 화면의 타이틀의 `JSON Data`를 찾아보니 "파리 만국박람회 1900(L'Exposition de Paris 1900)"와 같이 `Title`이 적혀있었습니다. 
- 그런데 기능 구현 영상에서 이를 찾아보니 `파리 만국박람회 1900` 부터 줄이 바뀌고, 그 다음줄부터 `(L'Exposition de Paris 1900)`가 적혀있었습니다. 이를 단순히 `lineBreakMode`를 어절 단위로 한다면 1900에서 끊기지 않고, `파리 만국 박람회`에서 먼저 끊기고, `1900(L'Exposition de Paris 1900)`이 그 다음 줄에 표시되게 됩니다. 
- 이는 구현 영상과는 포맷이 동일하지 않다고 판단하여, 1900의 뒤에 의도적으로 개행이 필요하다고 생각하였습니다.
- 이에 `replacingOccurrences` 메서드를 이용하여 1900 뒤의 열린 괄호 전에 아래와 같이 개행을 넣어주고, `numberOfLines`를 0으로 설정하니, 구현상황과 동일하게 개행이 이루어질 수 있었습니다.

```swift
private func updateTitleLabel() {
        guard let title = expositionUniverselle?.title else {
            return
        }
        
        let changeTitle = title.replacingOccurrences(of: "(", with: "\n(")
        
        titleLabel.text = changeTitle
        CustomLabel.setNumberOfLinesToZero(into: descriptionLabel)
        CustomLabel.setLabelFont(into: titleLabel, style: .title2)
}
```

- 구현상황과 동일하게 구현은 하였지만, 위와 같은 문제가 발생하였을 때, 더욱 좋은 방법으로 개행이 이루어질 수 있게 하는 방법이 있을지 궁금합니다.


<hr/>

이상 PR 마치겠습니다. 
이번에 고민한 것들이 많아서 내용이 다소 길어졌는데 읽어주셔서 감사합니다😁
이번 step2도 좋은 피드백 부탁드립니다 쿠마!🍀

yeeton37 committed:
feat: assets에 expo_assets 파일 추가

Jeon-Minsu committed:
feat: Main storyboard를 통한 만국박람회 초기 화면 구현

yeeton37 committed:
feat: IBOutlet 프로퍼티 생성 및 스토리보드와 dataAsset을 연결

Jeon-Minsu committed:
feat: ExpoError 파일 생성

Jeon-Minsu committed:
refactor: ViewController 코드를 메서드 단위로 분리 및 은닉화

yeeton37 committed:
feat: 버튼 클릭하면 두 번째 화면으로 넘어가는 segue 구현

Jeon-Minsu committed:
refactor: ViewController 내 navigationButton 클릭 시 화면 전환 코드 수정

Jeon-Minsu committed:
feat: ViewWillAppear 및 ViewWillDisAppear 호출시 네비게이션바 노출 여부 결정

yeeton37 committed:
feat: KoreaEntryViewController 생성 및 필수 메서드 구현

Jeon-Minsu committed:
refactor: ViewController 네이밍 변경 및 네비게이션 백 버튼 아이템 생성

Jeon-Minsu committed:
feat: KoreaEntryViewController 생성 및 필수 메서드 구현

Jeon-Minsu committed:
feat: KoreaEntryDetailsViewController 생성 및 IBOutlet 변수 내 값 할당

yeeton37 committed:
style: Assets 파일을 Resource 폴더로 이동시킴

yeeton37 committed:
refactor: KoreaEntryViewController 파일을 Controller 폴더로 이동 및 리팩토링

yeeton37 committed:
refactor: EXPOInformationViewController파일 Controller 폴더로 이동 및 리팩토링

yeeton37 committed:
refactor: KoreaEntryDetailsViewController파일 Controller 폴더로 이동 및 리팩토링

yeeton37 committed:
refactor: EXPOError파일 Model 폴더로 이동 및 리팩토링

yeeton37 committed:
feat: struct(JSONData) 선언

- KoreaEntryViewController, EXPOInformationViewController에서JSON data를 decode해주는 중복된 메서드를 하나의 struct(JSONData)로 만들어줌

yeeton37 committed:
feat: NavigationBarFormatter,CustomLabel 구조체 선언

- ViewController에서 중복해서 사용되는 UI 속성 변경 메서드를 구조체로 생성해줌

Jeon-Minsu committed:
docs: README.md 업데이트

yeeton37 committed:
docs: README.md 수정

yeeton37 committed:
docs: README.md 수정

yeeton37 committed:
docs: README.md 최종 수정

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -0,0 +1,121 @@
+class EXPOInformationViewController: UIViewController {
+    var expositionUniverselle: ExpositionUniverselle? = nil
+    
+    @IBOutlet weak var titleLabel: UILabel!
```
`access control` 을 고려해볼까요?

yeeton37 replied:
`private` 를 붙여주었습니다.!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    var expositionUniverselle: ExpositionUniverselle? = nil
+    
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var posterImage: UIImageView!
```
UI Component 를 명시해주면 좋을 것 같네요 : )
`posterImageView`

yeeton37 replied:
붙여주었습니다!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        self.navigationController?.isNavigationBarHidden = true
```
`self` 가 작성되는 곳이 있고, 생략하는 곳이 있습니다.
스타일을 하나로 맞춰주세요~!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+    
+    private func parseExpositionUniverselleData() {
+        guard let parsedInformation = JSONData.parse(name: "exposition_universelle_1900",
```
가급적이면 `string raw value` 는 구현부에서 피하면 어떨까요?
오타로 인한 에러나 유지보수에 악영향을 줄 수도 있습니다.

leejun6694 replied:
`name` 앞에서 줄바꿈을 할 수도 있습니다~!
한번 적용해보고, 더 마음에 드시는걸로 하시면 될 것 같아요!

Jeon-Minsu replied:
`NameSpace` 파일을 생성하여 열거형 내 케이스를 통해 `string raw value`를 구현부에서 사용하는 방식을 지양하였습니다.

Jeon-Minsu replied:
쿠마의 리뷰를 바탕으로 줄바꿈을 수행하니 더욱 코드 가독성이 높아졌습니다!

leejun6694 commented on Expo1900/Expo1900/Model/CustomLabel.swift:
```diff
@@ -0,0 +1,29 @@
+
+import UIKit
+
+struct CustomLabel {
```
```
extension UILabel {
}
```
을 활용해보는건 어떨까요?

yeeton37 replied:
`CustomLabel` 구조체를 지우고 
```swift
extension UILabel {

}
```
로 바꿔주었습니다!

leejun6694 commented on Expo1900/Expo1900/Model/CustomLabel.swift:
```diff
@@ -0,0 +1,29 @@
+
+struct CustomLabel {
+    static func setNumberOfLinesToZero(into label: UILabel) {
+        label.numberOfLines = 0
```
하나의 구현만 정의된 메서드를 따로 만드신 이유가 있을까요?
(아래 메서드들도..!)

Jeon-Minsu replied:
하나의 기능만 정의된 메서드를 사용하기 보다는, 파라미터에 만들고 싶은 `line`의 수를 입력받아, 이를 `label.numberOfLInes`로 할당하는 함수로 수정해보았습니다!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+    
+    private func updateDescriptionLabel() {
+        guard let description = expositionUniverselle?.description else {
```
같은 이유로 `guard` 문이 여러 곳에서 쓰이고 있는 것 같습니다.
이를 해결할 수 있을까요?

Jeon-Minsu replied:
옵셔널 바인딩이 필요한 이유가 `expositionUniverselle` 변수의 타입이 옵셔널이기 때문에, `guard`문이 반복적으로 사용된 것이었습니다. 이에, `expositionUniverselle`의 옵셔널 바인딩을 한 번 수행하여, 옵셔널 바인딩된 변수를 파라미터로 받아, 이를 바탕으로 `titleLabel`, `visitorsLabel`, `locationLabel`, `durationLabel`, `descriptionLabel` 의 값을 할당해주면, guard 문 반복사용을 줄일 수 있었습니다.

```swift
private func updatePosterDescription() {
        guard let expositionUniverselle = expositionUniverselle else {
            return
        }
        
        updateTitleLabel(from: expositionUniverselle)
        updateVisitorsLabel(from: expositionUniverselle)
        updateLocationLabel(from: expositionUniverselle)
        updateDurationLabel(from: expositionUniverselle)
        updateDescriptionLabel(from: expositionUniverselle)
}
```

leejun6694 replied:
네, 사실 상황에 따라서 구현이 달라지는 부분이지만, 한번 고민해보면 좋을 것 같아서 코멘트 드렸었습니다.
참고해주세요~!

leejun6694 commented on Expo1900/Expo1900/Controller/KoreaEntryDetailsViewController.swift:
```diff
@@ -0,0 +1,27 @@
+    var deliveredDescription: String = ""
+    var deliveredTitle: String = ""
+    
+    @IBOutlet weak private var entryImageView: UIImageView!
```
여기는 `access control` 이 적용되었군요 👍

leejun6694 commented on Expo1900/Expo1900/Controller/KoreaEntryDetailsViewController.swift:
```diff
@@ -0,0 +1,27 @@
+class KoreaEntryDetailsViewController: UIViewController {
+    var deliveredImageName: String = ""
+    var deliveredDescription: String = ""
+    var deliveredTitle: String = ""
```
하나의 모델로 만드는 것은 어떨까요!?

Jeon-Minsu replied:
`Exhibit`내 개별 프로퍼티인 `ImageName`, `description`, `title`을 보내기보다는, `Exhibits` 모델 그 자체를 전달하여 `KoreaEntryDetailsViewController`내에서 `Exhibits` 인스턴스의 프로퍼티에 접근하여 표현하는 방식으로 수정하였습니다.

leejun6694 commented on Expo1900/Expo1900/Controller/KoreaEntryViewController.swift:
```diff
@@ -0,0 +1,72 @@
+
+class KoreaEntryViewController: UIViewController {
+    @IBOutlet weak var tableView: UITableView!
+    var koreaEntry = [Exhibits]()
```
`@IBOutlet property` 가 먼저 정의되는 경우와 아닌 경우가 있습니다.
하나의 규칙으로 정리되면 좋을 것 같네요.

Jeon-Minsu replied:
변수가 먼저 정의되고, `@IBOutlet property`를 그 다음에 정의하는 방식으로 컨벤션을 통일하였습니다.

leejun6694 commented on Expo1900/Expo1900/Controller/KoreaEntryViewController.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+extension KoreaEntryViewController: UITableViewDataSource, UITableViewDelegate {
```
`extension` 을 사용해서 코드를 분리해주셨네요 💯

leejun6694 replied:
`UITableViewDelegate` 도 쓰이고 있나요?

Jeon-Minsu replied:
`UITableViewDelegate` 프로토콜 내 함수는 사용하지 않았기에, `extension`에서 제외하였습니다!

leejun6694 commented on Expo1900/Expo1900/DetailViewController.swift:
```diff
@@ -0,0 +1,31 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        // Do any additional setup after loading the view.
```
불필요한 코드는 pr 제출 시, 삭제 부탁드립니다 : )

yeeton37 replied:
삭제해주었습니다!

leejun6694 commented on Expo1900/Expo1900/DecodingError.swift:
```diff
@@ -0,0 +1,8 @@
+//  Created by 전민수 on 2022/06/15.
+//
+
+import Foundation
```
필요한 파일인가요?

Jeon-Minsu replied:
현재 `DecodingError`가 아닌 `EXPOError` 파일에서 에러를 핸들링하고 있습니다. `EXPOError`는 `extension`을 통해 `LocalizedError` 프로토콜을 채택하여, `NSLocalizedString`을 사용하기 때문에, `import Foundation`이 필요로 합니다.

yeeton37 replied:
저희가 파일을 지울 때 remove reference로 지워줘야 하는데 move to trash로 파일을 지워서 아직 남아있는 것 같습니다 ㅠㅠ 지워주었습니다!

leejun6694 commented on Expo1900/Expo1900/Model/NavigationBarFormatter.swift:
```diff
@@ -0,0 +1,17 @@
+//
+import UIKit
+
+struct NavigationBarFormatter {
```
요것도 `extension` 을 활용해볼까요?

yeeton37 replied:
넵!

leejun6694 reviewed (changes requested):
수꿍, yeton 안녕하세요 : )
코드를 깔끔하게 작성해주셔서, 읽는데 많은 도움이 되었습니다~!
몇 가지 리뷰를 남겼는데, 확인 부탁드려요.
하나의 리뷰로 여러 군데 적용되는 곳이 있으니, 참고 부탁드립니다.

yeeton37 committed:
refactor: enum(NameSpace) 생성 및 컨벤션 통일

yeeton37 committed:
refactor: struct(JSONData)의 메서드(Parse) 2개를 제너릭을 활용하여 1개로 통합

Jeon-Minsu committed:
refactor: 중복되는 guard 구문 수정, 컨벤션 통일 및 개별 프로퍼티가 아닌 모델(Exhibits) 전달

- expositionUniverselle 옵셔널 바인딩을 통한 중복되는 guard 구문 수정
- 각각의 ViewController는 더이상 상속을 받지 않는다 판단하여 class 앞에 final 키워드 추가
- 개별 프로퍼티인 deliveredImageName, deliveredDescription, deliveredTitle가 아닌 Exhibits 구조체의 인스턴스인 exhibit 전달
- UITableViewDelegate 프로토콜 미사용으로 인한 프로토콜 채택 제거
- parseKoreaEntryItems 메서드 줄바꿈 컨벤션 변경
- CustomLabel 구조체 생성이 아닌, UILabel extension 추가 및 파일명 변경
- UILabel extension 내 메서드에 파라미터를 추가하여, 재사용 및 유지보수 가능성 확장
- DecodingError 파일 삭제

yeeton37 committed:
refactor: NavigationBarFormatter을 extension UINavigationItem으로 변경 및 파일명 변경

Jeon-Minsu commented:
코멘트 바탕으로 전반적인 코드 수정 과정을 진행하였습니다!
의문점을 통해 해결하지 못한 부분을 제외하고는 모두 수정해보았습니다.
나머지 부분은 금일 22시에 같이 대화를 나누었으면 좋겠습니다!!

잘 부탁드립니다, 쿠마!! 🙇‍♂️🙇‍♀️

leejun6694 commented on Expo1900/Expo1900/Model/NameSpace.swift:
```diff
@@ -0,0 +1,53 @@
+//
+//  Created by 수꿍, yeton on 2022/06/20.
+//
+enum NameSpace: String {
```
오 이렇게까지 말씀드렸던 것은 아닌데,,
보여주는 값들은 굳이 이렇게 정리하지 않아도 될 것 같아요.
`identifier`, `file name` 등의 crash 나 구현에 문제가 생기는 부분들을 말씀드렸었습니다ㅎㅎ

Jeon-Minsu replied:
`crash` 혹은 구현 문제가 생기는 `string raw value`에 관하여서만 변수안에 집어 넣어 오타를 피하여야 함을 이해하였습니다. 이에, 함수 구현부에서 정의할 수 있으나, 현재 열거형 구조를 일단 만들어서 그 안에 수정을 진행해보았습니다. 오타가 발생하여도 파일에 문제가 생기지 않는 부분들은 열거형 케이스에서 제거해보았습니다.

leejun6694 commented on Expo1900/Expo1900/Model/NameSpace.swift:
```diff
@@ -0,0 +1,53 @@
+    case transferToDetailsVCId = "transferToDetailsVC"
+    case entryCellId = "entryCell"
+    
+    var name: String {
```
```
var name: String {
   return self.rawValue
}
```
참고해주세요~

Jeon-Minsu replied:
해당 방법을 참고하여 수정해보았습니다!

leejun6694 commented on Expo1900/Expo1900/Model/extension+UILabel.swift:
```diff
@@ -0,0 +1,29 @@
+
+extension UILabel {
+    func setNumberOfLines(_ line: Int) {
+        self.numberOfLines = line
```
이런거면,, 그냥 
`label.numberOfLines = 3`
요렇게 바로 써도 되지 않나요?
method 를 새로 정의한 이유가 궁금합니다 : )

Jeon-Minsu replied:
저희는 메서드를 통해 작성하는 것이 가독성을 더 높이고, 동료 개발자가 해당 코드가 어떻게 구현되어있는지 궁금할 때만, 해당 메서드를 살펴보면 되어 좋은 방법이라고 생각하여 진행했습니다.
혹시 쿠마는 이에 대하여 어떤 방법을 선호하시는지 질문드리고 싶습니다!

leejun6694 commented on Expo1900/Expo1900/Controller/KoreaEntryViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.tableView.dataSource = self
```
`tableView.delegate = self`
이것도 놓치지 않고 잘 지워주셨네요 👍

Jeon-Minsu replied:
감사합니다!

leejun6694 reviewed (approved):
리뷰 내용을 잘 반영해주셨네요~!
남은 의문점들은 내일 얘기해보아요~~!
고생하셨습니다 : )
